apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/transaction/types/smc/AbstractSmcTransactionType.java

```diff
@@ -0,0 +1,182 @@
+                               AccountService accountService,
+                               SmcContractRepository contractRepository,
+                               ContractToolService contractToolService,
+                               SmcFuelValidator fuelMinMaxValidator,
+                               SmcBlockchainIntegratorFactory integratorFactory,
+                               SmcConfig smcConfig) {
+        super(blockchainConfig, accountService);
+        this.contractRepository = contractRepository;
+        this.contractToolService = contractToolService;
+        this.fuelMinMaxValidator = fuelMinMaxValidator;
+        this.integratorFactory = integratorFactory;
+        this.blockchain = blockchain;
+        this.smcConfig = smcConfig;
+    }
+
+    @Override
+    public Fee getBaselineFee(Transaction transaction) {
+        //TODO calculate the required fuel value by executing the contract
+        //currently use: fee = fuelPrice * fuelLimit
+        AbstractSmcAttachment attachment = (AbstractSmcAttachment) transaction.getAttachment();
+        var fee = attachment.getFuelPrice().multiply(attachment.getFuelLimit()).longValueExact();
+        return new Fee.ConstantFee(fee);
+    }
+
+    @Override
+    public boolean canFailDuringExecution() {
+        return true;
+    }
+
+    @Override
+    public String getName() {
+        return getSpec().getCompatibleName();
+    }
+
+    @Override
+    public boolean applyAttachmentUnconfirmed(Transaction transaction, Account senderAccount) {
+        return true;
+    }
+
+    @Override
+    public void undoAttachmentUnconfirmed(Transaction transaction, Account senderAccount) {
+    }
+
+    @Override
+    public boolean isPhasingSafe() {
+        return false;
+    }
+
+    @Override
+    public boolean isPhasable() {
+        return false;
+    }
+
+    @Override
+    public boolean canHaveRecipient() {
+        return true;
+    }
+
+    @Override
+    public final void doStateIndependentValidation(Transaction transaction) throws AplUnacceptableTransactionValidationException {
+        checkPrecondition(transaction);
+        AbstractSmcAttachment attachment = (AbstractSmcAttachment) transaction.getAttachment();
+        fuelMinMaxValidator.validate(transaction);
+        executeStateIndependentValidation(transaction, attachment);
+    }
+
+    public abstract void executeStateIndependentValidation(Transaction transaction, AbstractSmcAttachment abstractSmcAttachment) throws AplUnacceptableTransactionValidationException;
+
+    private void checkPrecondition(Transaction smcTransaction) {
+        if (!getBlockchainConfig().isSmcTransactionsActiveAtHeight(blockchain.getHeight())) {
+            throw new AplTransactionFeatureNotEnabledException("'Smart-contract transactions'", smcTransaction);
+        }
+        smcTransaction.getAttachment().getTransactionTypeSpec();
+        if (smcTransaction.getAttachment().getTransactionTypeSpec() != getSpec()) {
```

Comment:
It is impossible and verified before this method call inside the AbstractAttachment.bindTransactionType

Reply:
removed